### PR TITLE
Danial/ede 338 make common problems public

### PR DESCRIPTION
**Pull request by danialmalik**
### Description
By default, edX allows public access to non-problem xBlocks. This PR adds the public access for Capa problems (Common Problems) provided by Open edX.

Message on course outline page for public courses is also updated
image

### How to test
- Make a course (or use an existing one) (https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/enable_public_course.html)
- Add problems in units
- Enable public access on the course
- Try to access and interact with problems while being an anonymous user.

### How to run unit tests
```
paver test_system -s lms -t lms/djangoapps/courseware/tests/test_module_render.py::ModuleRenderTestCase::test_anonymous_post_xblock_callback_with_public_access_enabled

paver test_system -s lms -t common/lib/xmodule/xmodule/tests/test_capa_module.py::CapaModuleTest::test_public_view

paver test_system -s lms -t common/lib/xmodule/xmodule/tests/test_capa_module.py::CapaModuleTest::test_submit_problem_with_authenticated_user

paver test_system -s lms -t common/lib/xmodule/xmodule/tests/test_capa_module.py::CapaModuleTest::test_submit_problem_with_anonymous_user

paver test_system -s lms -t common/lib/xmodule/xmodule/tests/test_capa_module.py::CapaModuleTest::test_save_problem_with_anonymous_user

paver test_system -s lms -t openedx/features/course_experience/tests/views/test_course_home.py::TestCourseHomePageAccess
```

### CODE REVIEW

File: common/lib/xmodule/xmodule/capa_base.py [hunk 1]
```diff
@@ -1463,6 +1482,32 @@ def save_problem(self, data):
         event_info['answers'] = answers
         _ = self.runtime.service(self, "i18n").ugettext
 
+        # Anonymous users cannot save problem.
+        if not self.runtime.user_id:
+            event_info['failure'] = 'anonymous_user'
+            self.track_function_unmask('save_problem_fail', event_info)
+            next_url = urlquote_plus(reverse('jump_to', kwargs={'course_id':self.location.course_key, 'location':self.location}))
+
+            regiteration_link = u'{lms_root}/register?next={next_url}'.format(
```
Comment (HamzaIbnFarooq) [inline common/lib/xmodule/xmodule/capa_base.py:1491]: Can we use reverse('register_user') here?

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 2]
```diff
@@ -578,6 +579,24 @@ def test_closed(self):
                                     due=self.yesterday_str)
         self.assertTrue(module.closed())
 
+
+    def test_public_view(self):
+        """
+        Test that public view for capa problems returns
```
Comment (HamzaIbnFarooq) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py:585]:
```suggestion
        Tests that public view for capa problems returns
```

File: common/lib/xmodule/xmodule/capa_base.py (hunk 1 above)
Comment (danialmalik) [reply common/lib/xmodule/xmodule/capa_base.py]: Done

File: common/lib/xmodule/xmodule/tests/test_capa_module.py (hunk 2 above)
Comment (danialmalik) [reply common/lib/xmodule/xmodule/tests/test_capa_module.py:588]: This is intentionally an imperative sentence.

**Comment by imhassantariq** [comment]
@danialmalik @tasawernawaz Tested this PR locally. Everything is working fine you can merge the changes.

File: common/lib/xmodule/xmodule/capa_base.py [hunk 3]
```diff
@@ -715,6 +726,12 @@ def get_problem_html(self, encapsulate=True, submit_notification=False):
                 u"Your answers were previously saved. Click '{button_name}' to grade them."
             ).format(button_name=self.submit_button_name())
 
+        # For anonymous users, don't show the attempts count.
+        if not self.runtime.user_id:
+            attempts_allowed = 0
+        else:
+            attempts_allowed = self.max_attempts
```
Comment (HamzaIbnFarooq) [inline common/lib/xmodule/xmodule/capa_base.py:733]:
```suggestion
        attempts_allowed = self.max_attempts if self.runtime.user_id else 0
```

**Pull request merged**